# K-Nearest Neighbors Regressor (KNN)

In [1]:
# 📌 Preparación de datos
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

#Carga de datos
from google.colab import files
uploaded = files.upload()
for fn in uploaded.keys():
    name=fn
df = pd.read_csv(name, sep=";", encoding='utf-8')

Saving Anexo ET_demo_round_traces_2022.csv to Anexo ET_demo_round_traces_2022.csv


<ipython-input-1-3ecd94279ebf>:14: DtypeWarning: Columns (6,7) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv(name, sep=";", encoding='utf-8')


In [2]:
# Codificación de variable categórica
le = LabelEncoder()
df['Team_encoded'] = le.fit_transform(df['Team'])

# Conversión a numérico
df['MatchKills'] = pd.to_numeric(df['MatchKills'], errors='coerce')
df['MatchAssists'] = pd.to_numeric(df['MatchAssists'], errors='coerce')
df['MatchHeadshots'] = pd.to_numeric(df['MatchHeadshots'], errors='coerce')
df['RoundKills'] = pd.to_numeric(df['RoundKills'], errors='coerce')

# Limpieza de datos nulos
df_clean = df.dropna(subset=['MatchKills', 'MatchAssists', 'MatchHeadshots', 'RoundKills', 'Team'])

# Variables
X = df_clean[['MatchAssists', 'MatchHeadshots', 'RoundKills', 'Team_encoded']]
y = df_clean['MatchKills']

from sklearn.svm import SVR
from sklearn.preprocessing import StandardScaler

# Escalado de datos
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

from sklearn.neighbors import KNeighborsRegressor

#Entrenamiento
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Modelo
model = KNeighborsRegressor(n_neighbors=5)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)

# Métricas
print("📎 K-Nearest Neighbors")
print(f"R²: {r2_score(y_test, y_pred):.3f}")
print(f"MAE: {mean_absolute_error(y_test, y_pred):.2f}")
print(f"MSE: {mean_squared_error(y_test, y_pred):.2f}")
print(f"RMSE: {np.sqrt(mean_squared_error(y_test, y_pred)):.2f}")
# División de datos

📎 K-Nearest Neighbors
R²: 0.700
MAE: 2.50
MSE: 11.47
RMSE: 3.39


# 📎 K-Nearest Neighbors (KNN)

Este modelo compara al jugador con otros jugadores similares y predice lo mismo que ellos.

Su R² fue el más bajo, con 0.700, y los errores más altos: MAE de 2.50 y RMSE de 3.39.

En la práctica, esto quiere decir que el modelo no logró distinguir bien las diferencias reales entre los jugadores, porque jugadores que se parecen en estadísticas no siempre se comportan igual en el juego. Por eso, su rendimiento fue el más débil